brainiak/utils/fmrisim.py

```diff
@@ -2134,126 +2701,229 @@ def generate_noise(dimensions,
+        as represented by the (desired) SFNR
+        - 'CNR_Amp/Noise-SD': Signal magnitude relative to the temporal
+        noise
+        - 'CNR_Amp2/Noise-Var_dB': Same as above but converted to decibels
+        - 'CNR_Signal-SD/Noise-SD': Standard deviation in signal
+        relative to standard deviation in noise
+        - 'CNR_Signal-Var/Noise-Var_dB': Same as above but converted to
+        decibels
+        - 'PSC': Calculate the percent signal change based on the
+        average activity of the noise (mean / 100 * magnitude)
 
-    percentile : float
-        What percentage of voxels will be included? Based on the values
-        supplied
 
     Returns
     ----------
-    ax : matplotlib object
-        Object with the information to be plotted
+    signal_function_scaled : 4d numpy array
+        The new signal volume with the appropriately set signal change
 
     """
 
-    ax = fig.add_subplot(111, projection='3d')
+    # If you have only one magnitude value, duplicate the magnitude for each
+    #  timecourse you have
+    if len(magnitude) == 1:
+        magnitude *= signal_function.shape[1]
 
-    # Threshold the data
-    threshold = np.percentile(brain.reshape(np.prod(brain.shape[0:3])),
-                              percentile)
+    # Scale all signals that to have a range of -1 to 1. This is
+    # so that any values less than this will be scaled appropriately
+    signal_function /= np.max(np.abs(signal_function))
 
-    # How many voxels exceed a threshold
-    brain_threshold = np.where(np.abs(brain) > threshold)
+    # Iterate through the timecourses and calculate the metric
+    signal_function_scaled = np.zeros(signal_function.shape)
+    for voxel_counter in range(signal_function.shape[1]):
 
-    # Clear the way
-    ax.clear()
+        # Pull out the values for this voxel
+        sig_voxel = signal_function[:, voxel_counter]
+        noise_voxel = noise_function[:, voxel_counter]
+        magnitude_voxel = magnitude[voxel_counter]
 
-    ax.set_xlim(0, brain.shape[0])
-    ax.set_ylim(0, brain.shape[1])
-    ax.set_zlim(0, brain.shape[2])
+        # Calculate the scaled time course using the specified method
+        if method == 'SFNR':
 
-    # If a mask is provided then plot this
-    if mask is not None:
-        mask_threshold = np.where(np.abs(mask) > 0)
-        ax.scatter(mask_threshold[0],
-                   mask_threshold[1],
-                   mask_threshold[2],
-                   zdir='z',
-                   c='black',
-                   s=10,
-                   alpha=0.01)
-
-    # Plot the volume
-    ax.scatter(brain_threshold[0],
-               brain_threshold[1],
-               brain_threshold[2],
-               zdir='z',
-               c='red',
-               s=20)
-
-    return ax
+            # How much temporal variation is there, relative to the mean
+            # activity
+            temporal_var = noise_voxel.mean() / noise_dict['sfnr']
```

Comment:
I think your concerns are completely valid but I am not sure of a better solution. You are right that when you measure SFNR you will account for temporal variability from both signal and noise, so adding signal on top of the noise that already exists will change the SFNR. However, I don't think that is necessarily important for three reasons: that effect will be negligible, since the SFNR for all voxels in the brain is used to calculate SFNR and presumably signal will only be added to a minuscule proportion; the amount of change to the SFNR will be within the range of normal changes to SFNR; and this metric is more useful when it only considers the SFNR of the data before signal is added, rather than after, because it simply quantifies the signal size relative to the temporal variability.

Reply:
I see. Then do you think SFNR is really critical to have? Can it be removed at all?
Also sorry I now get a bit confused: is SFNR prescribed in `noise_dict['sfnr']` or `magnitude_voxel`? If it is the former, what is the meaning of `magnitude_voxel` then?